### PR TITLE
WINC-1033: Use latest WICD binary during upgrades

### PR DESCRIPTION
**Pull request by saifshaikh48**
This PR enhances the node deconfiguration process to allow for better backward compatibility and failure recovery during upgrades.
- use the new version's WICD binary to cleanup during upgrade
- stop the WICD controller service from running before running WICD cleanup when recovering from node config failure

### CODE REVIEW

File: pkg/windows/windows.go [hunk 1]
```diff
@@ -404,12 +404,18 @@ func (vm *windows) RunWICDCleanup(apiServerURL string, watchNamespace string) er
...
 }
 
-func (vm *windows) Deconfigure(watchNamespace string, apiServerURL string) error {
+func (vm *windows) Deconfigure(watchNamespace string, apiServerURL string, credentials *Authentication) error {
 	vm.log.Info("deconfiguring")
 
 	if err := vm.deconfigureWICD(); err != nil {
 		return err
 	}
+	if err := vm.ensureWICDExists(credentials); err != nil {
```
Comment (sebsoto) [inline pkg/windows/windows.go:415]: Can you add a commit before this removing the `credentials` argument. it isnt used within ensureWICDExists

File: pkg/nodeconfig/nodeconfig.go [hunk 2]
```diff
@@ -190,7 +190,7 @@ func (nc *nodeConfig) Configure() error {
 			nodeConfigCache.credentials); err != nil {
 			return fmt.Errorf("configuring WICD failed: %w", err)
 		}
-		// Set the desired version annotation, communicating to WICD which Windows services configmap to use
+		// Set the desired version annotation
```
Comment (sebsoto) [inline pkg/nodeconfig/nodeconfig.go:193]: Why is this being changed?

File: pkg/daemon/cleanup/cleanup.go [hunk 3]
```diff
@@ -62,16 +62,23 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
+				return fmt.Errorf("node %s missing desired version annotation", node.Name)
+			}
 		}
 		err = directClient.Get(ctx,
-			client.ObjectKey{Namespace: configMapNamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+			client.ObjectKey{Namespace: configMapNamespace, Name: servicescm.NamePrefix + version}, cm)
 		if err != nil {
 			return err
 		}
+		klog.Infof("removing services tied to version: %s", version)
```
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:81]: nit: this is being logged to stdout and IMO will be noise when trying to debug the stdout/err stream on WMCO failures. At the least this should be debug level

File: pkg/windows/windows.go [hunk 4]
```diff
@@ -234,6 +234,8 @@ type Windows interface {
 	ConfigureWICD(string, string, *Authentication) error
 	// Deconfigure removes all files and networks created by WMCO and runs the WICD cleanup command.
 	Deconfigure(string, string, *Authentication) error
+	// DeconfigureWICD ensures the WICD service running on the Windows instance is removed
+	DeconfigureWICD() error
 	// RunWICDCleanup runs the WICD cleanup command that ensures all services configured by WICD are stopped.
 	RunWICDCleanup(string, string) error
```
Comment (sebsoto) [inline pkg/windows/windows.go:240]: I don't think we need two separate methods for this. Instead you can just ensure the service is stopped within  RunWICDCleanup.

**Review by sebsoto — CHANGES_REQUESTED** [review]

**Review by aravindhp — COMMENTED** [review]
@saifshaikh48 you are making multiple changes to the upgrade flow in a single PR and we don't have good upgrade coverage in CI. How can we ensure that this change does not break upgrades? Have you tested it out manually without using an upgrade simulation?

File: pkg/daemon/cleanup/cleanup.go (hunk 3 above)
Comment (saifshaikh48) [reply pkg/daemon/cleanup/cleanup.go:81]: I'm not seeing debug log capability with klog: https://github.com/kubernetes/klog/blob/v1.0.0/klog.go#L111

File: pkg/nodeconfig/nodeconfig.go (hunk 2 above)
Comment (saifshaikh48) [reply pkg/nodeconfig/nodeconfig.go:193]: I misunderstood its context (config flow vs deconfig flow), will revert

**Comment by alinaryan** [comment]
LGTM just curious about Aravindh's ask about testing

**Comment by saifshaikh48** [comment]
/test aws-e2e-upgrade

Hit a known flake in the deletion tests
```
timed out waiting for daemonset noop-ds to scale, current status: 
{CurrentNumberScheduled:2 NumberMisscheduled:0 DesiredNumberScheduled:2 NumberReady:1 ObservedGeneration:1 UpdatedNumberScheduled:2 NumberAvailable:1 NumberUnavailable:1 CollisionCount:<nil> Conditions:[]}
```

**Comment by aravindhp** [comment]
/retitle WINC-1033: Use latest WICD binary during upgrades

@saifshaikh48 lets try be succinct in our PR header. You can give more details within the PR.

File: pkg/windows/windows.go [hunk 5]
```diff
@@ -431,7 +431,7 @@ func (vm *windows) Bootstrap(desiredVer, apiServerURL, watchNamespace string, cr
 	}
 	// We have to ensure the WICD files exist separately before the rest of the files because we must ensure services
 	// are stopped and their files closed before modifying them.
-	if err := vm.ensureWICDExists(credentials); err != nil {
+	if err := vm.ensureWICDExists(); err != nil {
```
Comment (aravindhp) [inline pkg/windows/windows.go]: I know @sebsoto asked for this change but be aware if we ever decided to revert this PR, this orthogonal change which has nothing to do with this PR will get reverted too.

File: pkg/windows/windows.go [hunk 6]
```diff
@@ -233,9 +233,10 @@ type Windows interface {
 	// ConfigureWICD ensures that the Windows Instance Config Daemon is running on the node
 	ConfigureWICD(string, string, *Authentication) error
 	// Deconfigure removes all files and networks created by WMCO and runs the WICD cleanup command.
-	Deconfigure(string, string) error
+	Deconfigure(string, string, *Authentication) error
 	// RunWICDCleanup runs the WICD cleanup command that ensures all services configured by WICD are stopped.
-	RunWICDCleanup(string, string) error
+	// Stops the WICD service prior to executing cleanup, if needed.
```
Comment (aravindhp) [inline pkg/windows/windows.go:238]: `RunWICDCleanup ensures the WICD service  is stopped if needed and runs the cleanup command that ensures WICD managed services are stopped`

**Review by aravindhp — CHANGES_REQUESTED** [review]
Please call out why we need to use the new WICD binary in the `[windows] Stop WICD & copy new .exe before cleanup`. Nit: you can remove `.` from `.exe`

File: pkg/windows/windows.go (hunk 6 above)
Comment (saifshaikh48) [reply pkg/windows/windows.go]: updated

**Review by aravindhp — APPROVED** [review]

**Pull request merged**